packages/core/core-flows/src/definition/cart/steps/add-to-cart.ts

```diff
@@ -1,31 +1,103 @@
 import { ModuleRegistrationName } from "@medusajs/modules-sdk"
-import { CreateLineItemForCartDTO, ICartModuleService } from "@medusajs/types"
-import { StepResponse, createStep } from "@medusajs/workflows-sdk"
+import {
+  CartLineItemDTO,
+  CreateLineItemForCartDTO,
+  ICartModuleService,
+  UpdateLineItemWithSelectorDTO,
+} from "@medusajs/types"
+import { deepEqualObj, isPresent, MathBN } from "@medusajs/utils"
+import { createStep, StepResponse } from "@medusajs/workflows-sdk"
 
 interface StepInput {
+  id: string
   items: CreateLineItemForCartDTO[]
 }
 
 export const addToCartStepId = "add-to-cart-step"
 export const addToCartStep = createStep(
   addToCartStepId,
   async (data: StepInput, { container }) => {
-    const cartService = container.resolve<ICartModuleService>(
+    const cartModule = container.resolve<ICartModuleService>(
       ModuleRegistrationName.CART
     )
 
-    const items = await cartService.addLineItems(data.items)
+    const existingVariantItems = await cartModule.listLineItems({
+      cart_id: data.id,
+      variant_id: data.items.map((d) => d.variant_id!),
+    })
 
-    return new StepResponse(items, items)
+    const variantItemMap = new Map<string, CartLineItemDTO>(
+      existingVariantItems.map((item) => [item.variant_id!, item])
+    )
+
+    const itemsToCreate: CreateLineItemForCartDTO[] = []
+    const itemsToUpdate: UpdateLineItemWithSelectorDTO[] = []
+
+    for (const item of data.items) {
+      const existingItem = variantItemMap.get(item.variant_id!)
+      const metadataMatches =
+        (!isPresent(existingItem?.metadata) && !isPresent(item.metadata)) ||
+        deepEqualObj(existingItem?.metadata, item.metadata)
+
+      if (existingItem && metadataMatches) {
+        const quantity = MathBN.sum(
+          existingItem.quantity as number,
+          item.quantity || 1
+        ).toNumber()
+
+        itemsToUpdate.push({
+          selector: { id: existingItem.id },
+          data: { id: existingItem.id, quantity: quantity },
+        })
+      } else {
+        itemsToCreate.push(item)
+      }
+    }
+
+    const createdItems = itemsToCreate.length
+      ? await cartModule.addLineItems(itemsToCreate)
```

Comment:
do we have an upsert method?
"addLineItems" and "updateLineItems" should happen in 2 different steps to be able to revert.

Reply:
I believe it should work here since we are reverting both in the rollback no?

Reply:
if one of them fail we have a problem. for example the `add` works and `update` fails. we don't have anything on the compensation step as response.
Each step can only contain one operation with their equivalent compensation.

Reply:
yes thats true 👍

Reply:
Good catch, I've split the step into create and update in 2 steps 🤞🏻